Storage/SQL/BrowserDB.swift

```diff
@@ -206,6 +206,12 @@ open class BrowserDB {
         }
     }
 
+    func runQueryUnsafe<T>(_ sql: String, args: Args?, factory: @escaping (SDRow) -> T) -> Deferred<Maybe<Cursor<T>>> {
+        return withConnection { connection -> Cursor<T> in
+            connection.executeQueryUnsafe(sql, factory: factory, withArgs: args)
```

Comment:
You're leaking a cursor from inside the `withConnection` block. That cursor is live, and you'll be accessing it and finalizing it while other users are concurrently accessing the DB. Furthermore, you're passing this out through a `Deferred`, so you're passing a live cursor to, most likely, a different thread.

All of the work you do with a live cursor must happen inside the `withConnection` block.

Reply:
:+1:

Since this is already landed on master, I've opened up another PR — #3531 for this work.